optuna/samplers/_tpe/_truncnorm.py

```diff
@@ -0,0 +1,235 @@
+    right_hand_side = 1.0
+    numerator = 1.0
+    denom_factor = 1.0
+    denom_cons = 1 / a**2
+    sign = 1
+    i = 0
+
+    while abs(last_total - right_hand_side) > sys.float_info.epsilon:
+        i += 1
+        last_total = right_hand_side
+        sign = -sign
+        denom_factor *= denom_cons
+        numerator *= 2 * i - 1
+        right_hand_side += sign * numerator * denom_factor
+
+    return log_LHS + math.log(right_hand_side)
+
+
+def _log_ndtr(a: np.ndarray) -> np.ndarray:
+    return np.frompyfunc(_log_ndtr_single, 1, 1)(a).astype(float)
+
+
+def _norm_logpdf(x: np.ndarray) -> np.ndarray:
+    return -(x**2) / 2.0 - _norm_pdf_logC
+
+
+def _log_gauss_mass(a: np.ndarray, b: np.ndarray) -> np.ndarray:
+    """Log of Gaussian probability mass within an interval"""
+
+    # Calculations in right tail are inaccurate, so we'll exploit the
+    # symmetry and work only in the left tail
+    case_left = b <= 0
+    case_right = a > 0
+    case_central = ~(case_left | case_right)
+
+    def mass_case_left(a: np.ndarray, b: np.ndarray) -> np.ndarray:
+        return _log_diff(_log_ndtr(b), _log_ndtr(a))
+
+    def mass_case_right(a: np.ndarray, b: np.ndarray) -> np.ndarray:
+        return mass_case_left(-b, -a)
+
+    def mass_case_central(a: np.ndarray, b: np.ndarray) -> np.ndarray:
+        # Previously, this was implemented as:
+        # left_mass = mass_case_left(a, 0)
+        # right_mass = mass_case_right(0, b)
+        # return _log_sum(left_mass, right_mass)
+        # Catastrophic cancellation occurs as np.exp(log_mass) approaches 1.
+        # Correct for this with an alternative formulation.
+        # We're not concerned with underflow here: if only one term
+        # underflows, it was insignificant; if both terms underflow,
+        # the result can't accurately be represented in logspace anyway
+        # because sc.log1p(x) ~ x for small x.
+        return np.log1p(-_ndtr(a) - _ndtr(-b))
+
+    # _lazyselect not working; don't care to debug it
+    out = np.full_like(a, fill_value=np.nan, dtype=np.complex128)
+    if a[case_left].size:
+        out[case_left] = mass_case_left(a[case_left], b[case_left])
+    if a[case_right].size:
+        out[case_right] = mass_case_right(a[case_right], b[case_right])
+    if a[case_central].size:
+        out[case_central] = mass_case_central(a[case_central], b[case_central])
+    return np.real(out)  # discard ~0j
+
+
+def _bisect(f: Callable[[float], float], a: float, b: float, c: float) -> float:
+    if f(a) > c:
+        a, b = b, a
+    # TODO(amylase): Justify this constant
+    for _ in range(100):
+        m = (a + b) / 2
+        if f(m) < c:
+            a = m
+        else:
+            b = m
+    return m
+
+
+def _ndtri_exp_single(y: float) -> float:
+    # TODO(amylase): Justify this constant
+    return _bisect(_log_ndtr_single, -100, +100, y)
+
+
+def _ndtri_exp(y: np.ndarray) -> np.ndarray:
+    return np.frompyfunc(_ndtri_exp_single, 1, 1)(y).astype(float)
```

Comment:
```suggestion
def _ndtri_exp(z0: np.ndarray):
    C = 2 * math.sqrt(math.pi * 2)
    large = z0 > math.log(0.5)
    y = np.sqrt(-z0)
    y[large] = np.sqrt(-np.log(1 - np.exp(z0[large])))
    x = np.full_like(y, -1)
    for i in range(4):
        w = np.sqrt(-log_ndtr(x))
        x += C * (w - y) * w * np.exp(0.5 * x ** 2 - w ** 2)
    x[large] = -x[large]
    return x
```
[Possible followup] An algorithm based on Newton's method is simpler and should be a *lot* faster.